### ISSUE TITLE
[Bug] The Shell RemovePage function causes crash on iOS

### ISSUE THREAD

**Issue by ymebrugts** [xamarin/Xamarin.Forms#11218]
### Description
Removing pages from the navigation stack using RemovePage crashes on iOS

### Steps to Reproduce

1. GoToAsync() to a page a few times
2. The code below removes all the pages in the navigation stack. Except homepage because that one is not in this list. 

``` 
            var existingPages = Shell.Current.Navigation.NavigationStack.ToList();
            foreach (var page in existingPages)
            {
                if (page != null)
                {
                    Shell.Current.Navigation.RemovePage(page);
                }
            }
```

3. iOS TargetInvocationException when navigating or null reference exception when trying to go back.

### Expected Behavior
Navigation stack cleaned and you can go one page back to the home page. Any pages in between are removed.
### Actual Behavior
iOS crashes

### Basic Information

- Version with issue: 4,7 
  - iOS: 13

### Reproduction files
[RemovePageReproduction.zip](https://github.com/xamarin/Xamarin.Forms/files/4837051/RemovePageReproduction.zip)

**Comment by ymebrugts** [comment]
Reproduction was using 4.5 by mistake. New ones uploaded.

**Comment by samhouts** [comment]
When I run on Hot Restart, I get `Name: NSInvalidArgumentException Reason: -[Xamarin_Forms_Platform_iOS_ShellSectionRenderer_NavDelegate navigationController:animationControllerForOperation:fromViewController:toViewController:]: unrecognized selector sent to instance` on the first button press. Weird!

Anyway, I confirm the bug:

```
System.NullReferenceException: Object reference not set to an instance of an object
  at Xamarin.Forms.Shell.GetNavBarIsVisible (Xamarin.Forms.BindableObject obj) [0x00000] in D:\a\1\s\Xamarin.Forms.Core\Shell\Shell.cs:75 
  at Xamarin.Forms.Platform.iOS.ShellSectionRenderer+NavDelegate.WillShowViewController (UIKit.UINavigationController navigationController, UIKit.UIViewController viewController, System.Boolean animated) [0x00028] in D:\a\1\s\Xamarin.Forms.Platform.iOS\Renderers\ShellSectionRenderer.cs:522 
```

**Comment by pictos** [comment]
I took the liberty to investigate this issue, and maybe I found the solution. As you can see all the tests are green, so it's a good signal but I don't know if I broke some expected behavior.
![image](https://user-images.githubusercontent.com/20712372/85913550-3108cc80-b80c-11ea-8ec2-3a7db14dc0ae.png)

I realized that the behavior is different between `ShellNavigation` and `NavigationPage`. You can see the current behavior [in this video](https://youtu.be/z95Y5XhknyE)

**Comment by pictos** [comment]
cc @PureWeen

**Comment by PureWeen** [comment]
@pictos can you create a PR? The exception is coming from native iOS so the unit tests aren't really going to fail for this one AFAIK

**Comment by pictos** [comment]
@PureWeen sure thing. I will do it tomorrow morning.
Oh this is sad, I thought that tests covered my changes. Also I'll try to implement testing for this then

**Comment by PureWeen** [comment]
@pictos I started one here if you want to use it as a starting point

I hadn't really looked into the cause yet :-) 

```C#
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Xamarin.Forms.CustomAttributes;
using Xamarin.Forms.Internals;


#if UITEST
using Xamarin.UITest;
using NUnit.Framework;
using Xamarin.Forms.Core.UITests;
#endif

namespace Xamarin.Forms.Controls.Issues
{
	[Preserve(AllMembers = true)]
	[Issue(IssueTracker.Github, 11218, "[Bug] The Shell RemovePage function causes crash on iOS",
		PlatformAffected.iOS)]
#if UITEST
	[NUnit.Framework.Category(Core.UITests.UITestCategories.Github10000)]
	[NUnit.Framework.Category(UITestCategories.Shell)]
#endif
	public class Issue11218 : TestShell
	{
		protected override async void Init()
		{
			var page = AddContentPage();

			await Navigation.PushAsync(new ContentPage());

			await Task.Delay(5000);

			await Navigation.PushAsync(new LastPage());
		}

		public partial class LastPage : ContentPage
		{
			public LastPage()
			{
				var existingPages = Shell.Current.Navigation.NavigationStack.ToList();
				foreach (var page in existingPages)
				{
					if (page != null)
					{
						Shell.Current.Navigation.RemovePage(page);
					}
				}
			}
		}
	}
}
```

**Comment by pictos** [comment]
Thank you @PureWeen, I will use it for sure

**Comment by pictos** [comment]
@ymebrugts I saw your comment in the PR. And maybe this can take a while until getting merged, the team is very busy, I believe. So I did a workaround that you can try. I used it in your sample and works. You can find the snippet [here](https://gist.github.com/pictos/9009669761ceaf7096b38fd71a96e69e).

@PureWeen when I wrote this workaround I realize that we should rethink the `private` members inside classes that aren't `sealed`, I don't know if it makes sense, but I suggest to use `protected`, that way we can have easy access to the members when I inherit the class and do overrides. What do you think?

**Comment by ymebrugts** [comment]
> @ymebrugts I saw your comment in the PR. And maybe this can take a while until getting merged, the team is very busy, I believe. So I did a workaround that you can try. I used it in your sample and works. You can find the snippet [here](https://gist.github.com/pictos/9009669761ceaf7096b38fd71a96e69e).

How nice of you. I will try it out. Thank you very much.

**Comment by pictos** [comment]
@ymebrugts you're welcome, please let me know if it works.

**Comment by ymebrugts** [comment]
> @ymebrugts you're welcome, please let me know if it works.

It worked. Thanks :)

**Comment by xperseguers** [comment]
Just a side note... I see that this bug fix is targeted at the 5.0.0 milestone. Since it was reported for 4.7.0, why isn't it targeted at some 4.7.x instead? I mean this problem prevents me and many others logically to upgrade from 4.6 to 4.7. Now if it gets fixed in 5.0, it's unfortunately very likely that 4.8 up to 5.0 will introduce other bugs with the new features, so that's a bit problematic, isn't it?

**Comment by PureWeen** [comment]
I tested this on the latest 5.0 pre3 and it looks like it's fixed there

I would recommend using URI navigation to achieve what this PR here is trying to accomplish. 

The tricky part here is that the "RemovePage" call is happening the constructor of the page that's currently becoming visible so the new page isn't visible yet. This causes a bit of a weird visual effect as well

If you use URI navigation

If we're here => "\\page1\page2"

And then call this
```C#
GotoAsync("\\page1\page3")
```

Then shell will take care of removing page2 for you and on 5.0 it'll remove it smoothly
